Add TypeScript declaration file (index.d.ts)

lucasrmendonca opened the pull request:
This solves #12

lucasrmendonca committed:
Add TypeScript declaration file (index.d.ts)

lucasrmendonca committed:
Add TypeScript declaration file (index.d.ts)

eight04 commented on index.d.ts:
```diff
@@ -0,0 +1,27 @@
+export = launchWebAuthFlow;
+
+declare function launchWebAuthFlow(options: WebAuthFlowOptions): ResponseUrl;
```
Is there a way to tell typescript that this function returns a promise? i.e. this is an async function.

lucasrmendonca replied:
This can be done with `Promise<ResponseUrl>` I'll fix it in the next patch

eight04 commented on index.d.ts:
```diff
@@ -0,0 +1,27 @@
+  redirect_uri: string,
+  interactive?: boolean,
+  alwaysUseTab?: boolean,
+  windowOptions?: {
```
Is it possible to reuse type from other sources? Is there an official/3rd-party webextensions API type declaration that can be used?

lucasrmendonca replied:
Mozilla has an official auto-generated type declaration https://www.npmjs.com/package/@types/webextension-polyfill
however, I'm not sure how to create a type based on another type declaration, I'll investigate.

eight04 commented:
Is there a linter that works with `.d.ts` files similar to eslint? Since this file is written manually by hand, it may include errors.

lucasrmendonca commented:
> Is there a linter that works with `.d.ts` files similar to eslint? Since this file is written manually by hand, it may include errors.

There's [TSLint](https://palantir.github.io/tslint/) and we can add unit tests too, for extra safety.

lucasrmendonca commented:
New additions:
- Removed the "ResponseUrl" type since it was just an alias for type "string" anyways.
- Fix the return value from "string" to "Promise<string>"
- Re-use common types from "webextension-polyfill" (although we're importing the original module, TypeScript will fetch the official types from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/webextension-polyfill)
- Installed some TypeScript ESLint plugins and ran "npm run test" and fixed all linter suggestions (also modified the npm script to work for both *.js and *.ts files);

eight04 committed:
Fix: add ts to package

eight04 committed:
drop travis, switch to github action

eight04 commented on package.json:
```diff
@@ -15,15 +15,20 @@
     "@babel/core": "^7.11.1",
     "@babel/preset-env": "^7.11.0",
     "@rollup/plugin-babel": "^5.2.0",
+    "@types/webextension-polyfill": "^0.9.1",
```
Should we move this into prod-dependency? What is the common practice for this kind of peer-dependency type?

Or we should just document it in readme?

eight04 replied:
Or add a `"peerDependencies": {"types/webextension-polyfill": ">=0.9"}`?

lucasrmendonca replied:
hmm, it seems that the standard practice for TypeScript libraries is to move all imported `@types/**` dependencies into prod-dependency... however, I'm against bloating the dist for something that won't be used for all users, only those who are using TS.
What if we kept it as a devDependency and just documented it to the readme, like this?

Installing via npm:
`npm install webext-launch-web-auth-flow`

Installing via npm with TypeScript support:
`npm install webext-launch-web-auth-flow @types/webextension-polyfill`

?

eight04 committed:
Add: instruction to install types
